### PR TITLE
Add vl53l1x support

### PR DESCRIPTION
**Pull request by overheat**
## VL53L1X time of flight (TOF) sensor supporting
https://www.st.com/en/imaging-and-photonics-solutions/vl53l1x.html

### The vl53l1 library
To ease the usage of its vl53l1x time of flight sensor, STMicroelectronics provides this library, which is made of 2 parts :

- the core, that is generic to all platforms
- the platform, contains the adaptation layer, which will be implemented in the driver as it is Zephyr specific.

### The vl53l1x driver
Contains vl53l1x driver itself and the adaptation layer, which is implemented in the driver as it is Zephyr specific.

### The vl53l1x sample
vl53l1x  sensor sample, fetching proximity and distance data in every second. 

### Misc
documentation and overlay file

### CODE REVIEW

File: samples/sensor/vl53l1x/README.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
+.. _vl53l1x:
+
+VL53L1X: Time Of Flight sensor
+##############################
+
+Overview
+********
+This sample periodically measures distance between vl53l1x sensor
```
Comment (dbkinder) [inline samples/sensor/vl53l1x/README.rst]: measures the distance between the VL53L1X sensor and a target

File: samples/sensor/vl53l1x/README.rst [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+and target. The result is displayed on the console.
+It also shows how we can use the vl53l1x as a proximity sensor.
```
Comment (dbkinder) [inline samples/sensor/vl53l1x/README.rst]: VL53L1X

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: samples/sensor/vl53l1x/README.rst (hunk 2 above)
Comment (overheat) [reply samples/sensor/vl53l1x/README.rst]: Thank you! Already changed.

File: ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h [hunk 3]
```diff
@@ -0,0 +1,236 @@
+/*
+* Copyright (c) 2017, STMicroelectronics - All Rights Reserved
+*
+* This file is part of VL53L1 Core and is dual licensed,
+* either 'STMicroelectronics
+* Proprietary license'
+* or 'BSD 3-clause "New" or "Revised" License' , at your option.
```
Comment (MaureenHelm) [inline ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: Please review https://docs.zephyrproject.org/latest/contribute/index.html#contributing-non-apache-2-0-licensed-components and fill out the README template.

cc: @avisconti

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/vl53l1x/vl53l1_platform.c [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	if (status_int < 0) {
+		Status = VL53L1_ERROR_CONTROL_INTERFACE;
+		LOG_ERR("Failed to write");
+	}
+
+	return Status;
+}
+
+VL53L1_Error VL53L1_ReadMulti(VL53L1_DEV Dev, uint16_t index, uint8_t *pdata,
+			      uint32_t count)
```
Comment (avisconti) [inline drivers/sensor/vl53l1x/vl53l1_platform.c]: Same here.
What is the difference with VL53L0?
index is on 16-bit instead of 8-bit?

**Review by avisconti — CHANGES_REQUESTED** [review]
I'm not sure if it is possible to reuse some code for both VL53L0 and VL53L1 sensor. Nor if it is useful.
Maybe separating them is the right strategy. Pls add some comments to this.

If separation is required you may still want to adapt the copyright date.

File: drivers/sensor/vl53l1x/vl53l1_platform.c (hunk 4 above)
Comment (overheat) [reply drivers/sensor/vl53l1x/vl53l1_platform.c]: Yes, the VL53L1 index is on 16-bit instead of 8-bit

I have thought about re-use same driver but looks the internal implication is different between vl53l1x and vl53l0x, as you already have seen the "index"(register address) length is not the same. Also, those API related to interrupt and threshold and so on. So there should be lots of ifdef, if we want to merge those drivers together.

Comment (avisconti) [reply drivers/sensor/vl53l1x/vl53l1_platform.c]: > I have thought about re-use same driver but looks the internal implication is different between vl53l1x and vl53l0x, as you already have seen the "index"(register address) length is not the same. Also, those API related to interrupt and threshold and so on. So there should be lots of ifdef, if we want to merge those drivers together.

Yes, yes. In fact my initial idea was to re-use same driver, but the more I was going deep thru the review the less I found it applicable. So at the end I left it as a comment from your side.

File: ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h (hunk 3 above)
Comment (overheat) [reply ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: @avisconti  Thanks!
I have a question that those code under ext/hal/st/ comes from ST. Can I modify the copyright part as vl53l0x does?

Comment (avisconti) [reply ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: @overheat 
For the ext part yes. You may refer to the existing VL53L0X for reference.
For the driver/sensor/vl53l1x you may change/add ur name and 2019 in the copyright section.

I think you may need to change CODEOWNER to cover that code. You should be the owner for the driver, while I think that someone from ST should for what concern the ext part (maybe myself or @erwango)

Comment (overheat) [reply ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: @avisconti Thank you! I will update it. I will put you the CODEOWNER for the ext part, is that alright?

Comment (avisconti) [reply ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: @overheat 
OK, fine.

Comment (avisconti) [reply ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: @overheat 
I'm discussing with @erwango. I think that he can be the owner for it.

Comment (avisconti) [reply ext/hal/st/lib/sensor/vl53l1x/api/core/inc/vl53l1_api_calibration.h]: So, at the end I think that you won't need to change CODEOWNERS at all.

**Comment by overheat** [comment]
@avisconti @MaureenHelm @dbkinder  I have modified. Please review. Thanks!

**Comment by avisconti** [comment]
> @avisconti @MaureenHelm @dbkinder I have modified. Please review. Thanks!

Well, if I'm getting the change I guess this is not what we were expecting.
I mean, you do not have to change the ext code. The ext code should be left there unchanged, with the
original copyright message. At least this is what I have been told when I did a similar change.

You can take as an example the PR #14111 and especially the 52bc149 commit. You will see
the ext code unchanged from the original but with the following description in the commit msg:

commit 52bc149f0787e51a3ee44ad7833fec50cb0e34a1
Author: Armando Visconti <armando.visconti@st.com>
Date:   Tue Mar 5 15:56:18 2019 +0100

    ext/hal: Introduce STMEMS Standard C Drivers
    
    This package contains platform independent drivers written in C
    language for STMicroelectronics sensors. The aim of this package
    is to provide a common, clean and stable interface to access sensor
    registers.
    
    Library is located in ext/hal/st/stmemsc/
    
    Origin: ST Microelectronics
    License: BSD-3-Clause
    URL: https://www.st.com/en/embedded-software/c-driver-mems.html
    Commit: v1.00
    Purpose: provide a common and stable i/f to access sensor registers
    Maintained-by: ST Microelectronics
    
    Signed-off-by: Armando Visconti <armando.visconti@st.com>

Moreover, I also added a new README file: ext/hal/st/stmemsc/README (see again #14111)

So, to summarize, you have to keep the ext code original, then add a README file as I did, then
add the info in the commit msg as I did. Is it clear?

Instead, what I asked you to change is the copyright of the drivers/sensor/vl53l1x driver.
There you need to put 2019 as a date and your name.

Does it make sense to you?

**Review by dbkinder — APPROVED** [review]
Doc changes look OK, thanks.

**Comment by overheat** [comment]
@avisconti Thank you for your patience. I have restored the ext code back to  ST code original and add a README file, also modify the info in the commit msg. meanwhile, change the copyright of the drivers/sensor/vl53l1x driver.
Please help to review again, thanks a lot!

File: ext/hal/st/lib/sensor/vl53l1x/README [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+   In order to fit with Zephyr rules and simplify version updates, we have done a minimum of modifications :
+      - keep only API directory
+        (remove documentation and samples, see ST website for this)
+      - rename Api to api
+      - customize api/platform to fit in Zephyr (mainly for I2C access)
+        vl53l1x_platform.c and other platform related files have been moved to Zephyr drivers directory.
+
+   In order to use this library, you have to :
+      * define CONFIG_HAS_STLIB and CONFIG_VL53L1X
+      * use NEWLIB_LIBC in prj.conf (needed for abs function)
```
Comment (avisconti) [inline ext/hal/st/lib/sensor/vl53l1x/README]: I'm currently adding a similar external hal from ST Microelectronics for other sensors. I also need to
define NEWLIB_LIBC, but instead of letting the user to define it in prj.conf I used imply in Kconfig of the external
code. Have you considered to do the same?
See #14111 and in particular a40e62975 commit.

**Review by avisconti — CHANGES_REQUESTED** [review]
We are almost done. See my latest comment here about using 'imply' in Kconfig for NEWLIB_LIBC.
Moreover, can you squash together the related commits?
At the end I think that you should have only 3 commits:
  - ext: Add official ST library for vl53l1x
  - drivers: sensor: Add vl53l1x time of flight sensor driver
  - samples: sensor: Add vl53l1x time of flight sensor sample

Does it make sense?

File: ext/hal/st/lib/sensor/vl53l1x/README (hunk 5 above)
Comment (overheat) [reply ext/hal/st/lib/sensor/vl53l1x/README]: yes, sound better. 
wait a moment, I will push again.

Comment (avisconti) [reply ext/hal/st/lib/sensor/vl53l1x/README]: OK, just remember to squash together similar commits as I was saying before.
I suggest to save the branch before squashing just in case

**Review by avisconti — APPROVED** [review]
Good job!

**Comment by overheat** [comment]
Fix mirror bugs.

**Comment by avisconti** [comment]
@overheat 
Pls take a look at #15857. It should fix the CI error you are having here.

**Comment by overheat** [comment]
>@overheat  ...

@avisconti yes, should I wait until that PR merged then re-trigger this CI check again?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by avisconti** [comment]
> @avisconti yes, should I wait until that PR merged then re-trigger this CI check again?

Maybe you can get my commit (c18747f) from my remote repo on github, rebase on top of it and re-push everything:

-----
git add remote avisconti 	https://github.com/avisconti/zephyr.git
git fetch avisconti
git cherry-pick c18747f
git push -f ....
-----

I think it is important to know if it is fixing or not.

**Comment by overheat** [comment]
> > @avisconti yes, should I wait until that PR merged then re-trigger this CI check again?
> 
> Maybe you can get my commit ([c18747f](https://github.com/zephyrproject-rtos/zephyr/commit/c18747faf548d8743e2f8d6d3d62cdefbca39a16)) from my remote repo on github, rebase on top of it and re-push everything:
> 
> git add remote avisconti https://github.com/avisconti/zephyr.git
> git fetch avisconti
> git cherry-pick [c18747f](https://github.com/zephyrproject-rtos/zephyr/commit/c18747faf548d8743e2f8d6d3d62cdefbca39a16)
> git push -f ....
> 
> I think it is important to know if it is fixing or not.

Done, let's see what happen.

**Comment by overheat** [comment]
@avisconti it works. I will push again when #15857 be merged.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by avisconti** [comment]
superseeded by #18772
